apps/arweave/src/ar_mining_worker.erl

```diff
@@ -0,0 +1,535 @@
+	#mining_candidate{ chunk2 = Chunk2 } = Candidate,
+	case cycle_chunk_cache(Candidate, {chunk2, Chunk2}, Map) of
+		{{chunk1, Chunk1, H1}, Map2} ->
+			%% Decrement 2 for chunk1 and chunk2:
+			%% 1. chunk1 was previously read and cached
+			%% 2. chunk2 that was just read and will shortly be used to compute h2
+			ar_mining_server:update_cache_usage(-2),
+			ar_mining_hash:compute_h2(
+				self(), Candidate#mining_candidate{ chunk1 = Chunk1, h1 = H1 }),
+			Session2 = Session#mining_session{ chunk_cache = Map2 },
+			{noreply, State#state{ session = Session2 }};
+		{{chunk1, H1}, Map2} ->
+			%% Decrement 1 for chunk2:
+			%% we're computing h2 for a peer so chunk1 was not previously read or cached 
+			%% on this node
+			ar_mining_server:update_cache_usage(-1),
+			ar_mining_hash:compute_h2(self(), Candidate#mining_candidate{ h1 = H1 }),
+			Session2 = Session#mining_session{ chunk_cache = Map2 },
+			{noreply, State#state{ session = Session2 }};
+		{do_not_cache, Map2} ->
+			%% Decrement 1 for chunk2
+			%% do_not_cache indicates chunk1 was not and will not be read or cached
+			ar_mining_server:update_cache_usage(-1),
+			Session2 = Session#mining_session{ chunk_cache = Map2 },
+			{noreply, State#state{ session = Session2 }};
+		{cached, Map2} ->
+			Session2 = Session#mining_session{ chunk_cache = Map2 },
+			{noreply, State#state{ session = Session2 }}
+	end;
+
+handle_task({compute_h0, Candidate}, State) ->
+	#state{ session = #mining_session{ seed = Seed } } = State,
+	ar_mining_hash:compute_h0(
+		self(),
+		Candidate#mining_candidate{
+			seed = Seed
+		}),
+	{noreply, State};
+
+handle_task({computed_h0, Candidate}, State) ->
+	case ar_mining_server:has_cache_space() of
+		true ->
+			#mining_candidate{ h0 = H0, partition_number = PartitionNumber,
+				partition_upper_bound = PartitionUpperBound } = Candidate,
+			{RecallRange1Start, RecallRange2Start} = ar_block:get_recall_range(H0,
+					PartitionNumber, PartitionUpperBound),
+			PartitionNumber2 = ?PARTITION_NUMBER(RecallRange2Start),
+			Candidate2 = Candidate#mining_candidate{ partition_number2 = PartitionNumber2 },
+			Candidate3 = generate_cache_ref(Candidate2),
+			Range1Exists = ar_mining_io:read_recall_range(
+					chunk1, self(), Candidate3, RecallRange1Start),
+			case Range1Exists of
+				true ->
+					ar_mining_server:reserve_cache_space(),
+					Range2Exists = ar_mining_io:read_recall_range(
+							chunk2, self(), Candidate3, RecallRange2Start),
+					case Range2Exists of
+						true -> ar_mining_server:reserve_cache_space();
+						false -> signal_cache_cleanup(Candidate3)
+					end;
+				false ->
+					?LOG_DEBUG([{event, mining_debug_no_io_thread_found_for_range},
+						{worker, State#state.name},
+						{range_start, RecallRange1Start},
+						{range_end, RecallRange1Start + ?RECALL_RANGE_SIZE}]),
+					ok
+			end;
+		false ->
+			%% Re-add the task so that it can bd executed later once some cache space frees up.
+			add_task(self(), computed_h0, Candidate)
```

Comment:
Maybe we should fix it in future.
1. Limit pending h0 tasks
Because we need to check how many h0 tasks are already in queue. So if this task is oldest and we hit some limit on h0 tasks, so we don't add it because we are overloaded
2. Decrease priority to prevent CPU waste
If queue is already filled with lots of tasks dry-run pull/push is not kinda good thing. So if this task was pulled with some priority and we can't handle it right now, we should push it with reduced priority so ensure we will not pull this same task again and waste CPU one more time

Reply:
the tasks are currently sorted based on task type and step number. `compute_h0` is the lowest priority, `computed_h0` is the next lowest. Within a task type the older the step number the lower the priority. So this task is already likely to be the highest priority pending task (typically the h1/h2 and chunk1/chunk2 tasks get handled immediately, and it's only the compute_h0 and computed_h0 that can get queued)

Also every time there's a new VDF session, any pending tasks for the retired VDF session are purged from the queue.

I agree the pop/push behavior isn't ideal, but I think it's also negligible in terms of performance. In my investigation it wasn/t the gb_sets operations that were slow, but the work that happens when a task is actually executed (and even that was only a problem at high partition counts for a single thread - so the new parallelized task queue should address all the task management bottlenecks that I've seen)